Humongous search terms in generated algorithm in mapping service

### How to Reproduce
Upload:
[lifecycle_test.xlsx](https://github.com/molgenis/molgenis/files/2072425/lifecycle_test.xlsx)
Create mapping for harmonizations, map from the cohort harmonizations

### Expected behavior
Mapping succeeds

### Observed behavior
```
2018-06-05_14:35:34.784 [pool-3-thread-1] ERROR o.m.data.index.job.IndexJobService - Index job failed
java.lang.IllegalArgumentException: Document contains at least one immense term in field="algorithm_fdfc208e.raw" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[36, 40, 39, 116, 97, 114, 103, 101, 116, 39, 41, 46, 109, 97, 112, 40, 123, 34, 97, 103, 101, 98, 105, 114, 116, 104, 95, 109, 95, 100]...', original message: bytes can be at most 32766 in length; got 36734
	at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:806)
	at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:447)
	at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:403)
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:232)
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:478)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1569)
	at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1314)
	at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:639)
	at org.elasticsearch.index.engine.InternalEngine.indexIntoLucene(InternalEngine.java:583)
	at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:483)
	at org.elasticsearch.index.shard.IndexShard.index(IndexShard.java:560)
	at org.elasticsearch.index.shard.IndexShard.index(IndexShard.java:549)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.executeIndexRequestOnPrimary(TransportShardBulkAction.java:484)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:143)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:113)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:69)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:939)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:908)
	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:113)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.onResponse(TransportReplicationAction.java:322)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.onResponse(TransportReplicationAction.java:264)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$1.onResponse(TransportReplicationAction.java:888)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$1.onResponse(TransportReplicationAction.java:885)
	at org.elasticsearch.index.shard.IndexShardOperationsLock.acquire(IndexShardOperationsLock.java:147)
	at org.elasticsearch.index.shard.IndexShard.acquirePrimaryOperationLock(IndexShard.java:1656)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.acquirePrimaryShardReference(TransportReplicationAction.java:897)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.access$400(TransportReplicationAction.java:93)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:281)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:260)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:252)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69)
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:627)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: max_bytes_length_exceeded_exception: bytes can be at most 32766 in length; got 36734
	at org.apache.lucene.util.BytesRefHash.add(BytesRefHash.java:263)
	at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:149)
	at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:796)
	... 37 common frames omitted
```

--- a/molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/client/MappingContentBuilder.java
+++ b/molgenis-data-elasticsearch/src/main/java/org/molgenis/data/elasticsearch/client/MappingContentBuilder.java
@@ -19,6 +19,12 @@ import static org.molgenis.data.elasticsearch.FieldConstants.FIELD_NOT_ANALYZED;
  */
 class MappingContentBuilder
 {
+	/**
+	 * Protect against Lucene’s term byte-length limit of 32766 bytes
+	 * 32766/4=8191 since UTF-8 characters may occupy at most 4 bytes
+	 */
+	private static final int IGNORE_ABOVE_VALUE = 8191;
+
 	private final XContentType xContentType;
 
 	MappingContentBuilder()
@@ -147,7 +153,7 @@ class MappingContentBuilder
 		XContentBuilder fieldsObject = contentBuilder.startObject("fields")
 													 .startObject(FIELD_NOT_ANALYZED)
 													 .field("type", "keyword")
-													 .field("index", true)
+													 .field("index", true).field("ignore_above", IGNORE_ABOVE_VALUE)
 													 .endObject();
 		fieldsObject.endObject();
 	}

--- a/molgenis-data-elasticsearch/src/test/java/org/molgenis/data/elasticsearch/client/MappingContentBuilderTest.java
+++ b/molgenis-data-elasticsearch/src/test/java/org/molgenis/data/elasticsearch/client/MappingContentBuilderTest.java
@@ -76,6 +76,6 @@ public class MappingContentBuilderTest
 	private static final String JSON_DOUBLE = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"double\"}}}";
 	private static final String JSON_INTEGER = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"integer\",\"doc_values\":true}}}";
 	private static final String JSON_LONG = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"long\"}}}";
-	private static final String JSON_TEXT = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"text\",\"norms\":true,\"fields\":{\"raw\":{\"type\":\"keyword\",\"index\":true}}}}}";
+	private static final String JSON_TEXT = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"text\",\"norms\":true,\"fields\":{\"raw\":{\"type\":\"keyword\",\"index\":true,\"ignore_above\":8191}}}}}";
 	private static final String JSON_NESTED = "{\"_source\":{\"enabled\":false},\"properties\":{\"field\":{\"type\":\"nested\",\"properties\":{\"nestedField\":{\"type\":\"boolean\"}}}}}";
 }

